### PR TITLE
[CDAP-1369] Create/delete namespace directories on namespace create/dele...

### PR DESCRIPTION
**Pull request by bdmogal**
...te. Re-org CDAP directory structure

Summary:
1. Creates ``<cdap_root>/<namespace-id>`` directory on namespace create
2. Deletes ``<cdap_root>/<namespace-id>`` directory on namespace delete
3. Underlying directory structure:
```
<cdap_root>/<namespace-id>
<cdap_root>/<namespace-id>/${app.output.dir}
<cdap_root>/<namespace-id>/${app.output.dir}/PurchaseHistory
<cdap_root>/<namespace-id>/${app.output.dir}/PurchaseHistory/Flow/PurchaseFlow.jar
<cdap_root>/<namespace-id>/${app.output.dir}/PurchaseHistory/Mapreduce/PurchaseHistoryBuilder.jar
<cdap_root>/<namespace-id>/${app.output.dir}/PurchaseHistory/Service/CatalogLookup.jar
<cdap_root>/<namespace-id>/${app.output.dir}/PurchaseHistory/Workflow/PurchaseHistoryWorkflow.jar
<cdap_root>/<namespace-id>/${app.output.dir}/PurchaseHistory/archive/Purchase-2.8.0-SNAPSHOT.jar
<cdap_root>/<namespace-id>/tmp:
```

There will be follow up PRs for similar directory structure changes for logs, datasets, streams and system lib directories.

### CODE REVIEW

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/AppFabricTestBase.java [hunk 1]
```diff
@@ -58,8 +58,11 @@
 import org.apache.http.util.EntityUtils;
 import org.apache.twill.discovery.DiscoveryServiceClient;
 import org.apache.twill.discovery.ServiceDiscovered;
+import org.apache.twill.filesystem.Location;
```
Comment (anwar6953) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/AppFabricTestBase.java]: can remove these imports?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java [hunk 2]
```diff
@@ -328,7 +326,17 @@ public void stop(Id.Namespace id, Id.Program programId, ProgramType type) throws
...
-      Location destination = locationFactory.create(archiveDir).append(namespaceId).append(adapterTypeInfo.getType());
+      Location namespaceHomeLocation = locationFactory.create(namespaceId);
+      if (!namespaceHomeLocation.exists()) {
+        String msg = String.format("Home directory %s for namespace %s not found",
+                                   namespaceHomeLocation.toURI().getPath(), namespaceId);
+        LOG.error(msg);
+        throw new FileNotFoundException(msg);
+      }
+
+      String appFabricDir = configuration.get(Constants.AppFabric.OUTPUT_DIR);
```
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java]: rename it? `appFabricDir` doesn't seem appropriate. Perhaps `archiveDir`

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/AppFabricTestBase.java (hunk 1 above)
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/AppFabricTestBase.java]: will check again

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java (hunk 2 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java]: well, ``archiveDir = appFabricDir + "/archive"``.

Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java]: oh ok, I see what you're saying.

**Comment by bdmogal** [comment]
Build: http://builds.cask.co/browse/CDAP-DUT743-2

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java [hunk 3]
```diff
@@ -120,11 +131,29 @@ public boolean hasNamespace(Id.Namespace namespaceId) {
...
+    // TODO: CDAP-1427 - This should be transactional, but we don't support transactions on files yet
     NamespaceMeta existing = store.createNamespace(metadata);
     if (existing != null) {
       throw new AlreadyExistsException(NAMESPACE_ELEMENT_TYPE, metadata.getId());
     }
+    String namespaceId = metadata.getId();
+    Location namespaceLocation = locationFactory.create(metadata.getId());
+    String namespacePath = namespaceLocation.toURI().getPath();
+    if (namespaceLocation.exists()) {
+      // need this hack because default namespace cannot be deleted, so the directory is always found, once created
```
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java]: I believe that this problem will still persist even when default_namespace creation is moved to a more deterministic time.

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/AppFabricTestBase.java (hunk 1 above)
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/AppFabricTestBase.java]: Fixed

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java (hunk 3 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java]: Yes, you are right. Fixing it.

**Comment by anwar6953** [comment]
LGTM

**Comment by bdmogal** [comment]
Build succeeded - http://builds.cask.co/browse/CDAP-DUT743-5. Merging.

**Pull request merged**